src/Features/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexEmbeddedCompletionProvider.cs

```diff
@@ -0,0 +1,546 @@
+
+        private readonly RegexEmbeddedLanguageFeatures _language;
+
+        public RegexEmbeddedCompletionProvider(RegexEmbeddedLanguageFeatures language)
+        {
+            _language = language;
+        }
+
+        public override bool ShouldTriggerCompletion(SourceText text, int caretPosition, CompletionTrigger trigger, OptionSet options)
+        {
+            if (trigger.Kind == CompletionTriggerKind.Invoke ||
+                trigger.Kind == CompletionTriggerKind.InvokeAndCommitIfUnique)
+            {
+                return true;
+            }
+
+            if (trigger.Kind == CompletionTriggerKind.Insertion)
+            {
+                return IsTriggerCharacter(trigger.Character);
+            }
+
+            return false;
+        }
+
+        private bool IsTriggerCharacter(char ch)
+        {
+            switch (ch)
+            {
+                case '\\': // any escape
+                case '[':  // character class
+                case '(':  // any group
+                case '{':  // \p{
+                    return true;
+            }
+
+            return false;
+        }
+
+        public override async Task ProvideCompletionsAsync(CompletionContext context)
+        {
+            var embeddedContext = new EmbeddedCompletionContext(context);
+            await ProvideCompletionsAsync(embeddedContext).ConfigureAwait(false);
+
+            if (embeddedContext.Items.Count == 0)
+            {
+                return;
+            }
+
+            var index = 0;
+            foreach (var embeddedItem in embeddedContext.Items)
+            {
+                var change = embeddedItem.Change;
+                var textChange = change.TextChange;
+
+                var properties = ImmutableDictionary.CreateBuilder<string, string>();
+                properties.Add(StartKey, textChange.Span.Start.ToString());
+                properties.Add(LengthKey, textChange.Span.Length.ToString());
+                properties.Add(NewTextKey, textChange.NewText);
+                properties.Add(DescriptionKey, embeddedItem.FullDescription);
+                properties.Add(AbstractEmbeddedLanguageCompletionProvider.EmbeddedProviderName, this.Name);
+
+                if (change.NewPosition != null)
+                {
+                    properties.Add(NewPositionKey, change.NewPosition.ToString());
+                }
+
+                // Keep everything sorted in the order we just produced the items in.
+                var sortText = index.ToString("0000");
+
+                var item = CompletionItem.Create(
+                    displayText: embeddedItem.DisplayText,
+                    sortText: sortText,
+                    properties: properties.ToImmutable(),
+                    rules: s_rules,
+                    inlineDescription: "    " + embeddedItem.InlineDescription);
```

Comment:
note the `"    "` can go away once the editor right-aligns completion-item suffixes.